src/init/features/genkit/index.ts

```diff
@@ -0,0 +1,581 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fs from "fs";
+import * as inquirer from "inquirer";
+import * as path from "path";
+import * as semver from "semver";
+import { execFileSync } from "child_process";
+
+import { doSetup as functionsSetup } from "../functions";
+import { Config } from "../../../config";
+import { promptOnce } from "../../../prompt";
+import { wrapSpawn } from "../../spawn";
+import { Options } from "../../../options";
+import { getProjectId } from "../../../projectUtils";
+import { ensure } from "../../../ensureApiEnabled";
+import { logger } from "../../../logger";
+import { FirebaseError } from "../../../error";
+
+interface GenkitInfo {
+  genkitVersion: string;
+  sampleVersion: string;
+  useInit: boolean;
+}
+
+function getGenkitVersion(): GenkitInfo {
+  let genkitVersion: string;
+  let sampleVersion: string | undefined;
+  let useInit: boolean;
+
+  // Allow the installed version to be set for dev purposes.
+  if (process.env.GENKIT_DEV_VERSION && typeof process.env.GENKIT_DEV_VERSION === "string") {
+    semver.parse(process.env.GENKIT_DEV_VERSION);
+    genkitVersion = process.env.GENKIT_DEV_VERSION;
+  } else {
+    genkitVersion = execFileSync("npm", ["view", "genkit", "version"]).toString();
+  }
+
+  if (!genkitVersion) {
+    throw new FirebaseError("Unable to determine genkit version to install");
+  }
+
+  if (semver.gte(genkitVersion, "1.0.0")) {
+    // We don't know about this version.
+    throw new FirebaseError(
+      "Please update your firebase-tools. Alternatively you can set a GENKIT_DEV_VERSION environment variable to choose a genkit version < 1.0.0",
+    );
+  } else if (semver.gte(genkitVersion, "0.6.0")) {
+    sampleVersion = "0.9.0";
+    useInit = false;
+  } else {
+    sampleVersion = "";
+    useInit = true;
+  }
```

Comment:
The whole reason we're doing this PR is because we don't want things to break and have no solution.
With this setup, I can submit this PR now, people can update to it, and it will continue to allow them to install 0.5.17 (the latest) up until we release 0.9.0 and then it will install 0.9.0. 
If I don't do this, then I can't submit this PR until sometime after we release 0.9.0 and there will be a period where it is just broken with no "just update to the latest firebase-tools" fix. 
It would also not notify people to update their firebase-tools because a newer version (>=1.0.0) is available. 
If we had this kind of logic in place from the beginning, then nothing would break when we release 0.9.0. So I'm a bit confused why you want me to remove it. People do not necessarily update their firebase-tools unless they have a reason to do so. 
I wouldn't want them to be "stuck" getting `firebase init genkit` installs for 0.9.0 when the latest version of genkit is 1.5.7 and there have been breaking change and the docs are for 1.5.7 and they just have a bad time trying to figure out what's going on and end up not liking genkit because it's "broken". I'd much rather have them getting a message saying "you need to update your firebase-tools, or explicitly pick an "old" version of genkit to install". If we do it this way then, we can remove support for older versions and add support for newer versions and not be coupled to the release of genkit versions without breaking anything. I feel very strongly that this is a good solution.

Reply:
support 0.5 is fine... 
I'm more thinking about the `semver.gte(genkitVersion, "1.0.0")` case. I think that if version is `semver.gte(genkitVersion, "0.6.0")` we pin it to 0.9. until we update the firebase cli in the future.